content/docs/rfcs/36/README.md

```diff
@@ -0,0 +1,738 @@
+2. `char* pubsubTopic`: Pubsub topic to verify.
+   If `NULL`, it verifies the number of peers in the default pubsub topic.
+
+**Returns**
+
+A [`JsonResponse`](#jsonresponse).
+If the execution is successful, the `result` field contains a `boolean` indicating whether there are enough peers.
+
+For example:
+
+```json
+{
+  "result": true
+}
+```
+
+### `extern char* waku_relay_subscribe(int nodeId, char* pubsubTopic)`
+
+Subscribe to a Waku Relay pubsub topic to receive messages.
+
+**Parameters**
+
+1. `int nodeId`: The node identifier obtained from a successful execution of [`waku_new`](#extern-char-waku_newchar-configjson).
+2. `char* pubsubTopic`: Pubsub topic to subscribe to. 
+   If `NULL`, it subscribes to the default pubsub topic.
+
+**Returns**
+
+A [`JsonResponse`](#jsonresponse).
+If the execution is successful, the `result` field contains an `integer` that is the subscription id.
+
+For example:
+
+```json
+{
+  "result": 0
+}
+```
+
+**Events**
+
+When a message is received, a ``"message"` event` is emitted containing the message, pubsub topic, and node ID in which
+the message was received.
+
+The `event` type is [`JsonMessageEvent`](#jsonmessageevent).
+
+For Example:
+
+```json
+{
+  "nodeId": 1,
+  "type": "message",
+  "event": {
+    "subscriptionID": 1,
+    "pubsubTopic": "/waku/2/default-waku/proto",
+    "messageID": "0x6496491e40dbe0b6c3a2198c2426b16301688a2daebc4f57ad7706115eac3ad1",
+    "wakuMessage": {
+      "payload": "TODO",
+      "contentTopic": "/my-app/1/notification/proto",
+      "version": 1,
+      "timestamp": 1647826358000000000
+    }
+  }
+}
+```
+
+### `extern char* waku_relay_close_subscription(int nodeId, char* subscriptionId)`
+
+Closes a Waku Relay subscription.
+No more messages will be received from this subscription.
+
+**Parameters**
+
+1. `int nodeId`: The node identifier obtained from a successful execution of [`waku_new`](#extern-char-waku_newchar-configjson).
+2. `char* subscriptionId`: Subscription ID to close.
+
+**Returns**
+
+A [`JsonResponse`](#jsonresponse).
+If the execution is successful, the `result` field is set to `null`.
+
+For example:
+
+```json
+{
+   "result": null
+}
+```
+
+### `extern char* waku_relay_unsubscribe_from_topic(int nodeId, char* topic)`
+
+Closes the pubsub subscription to a pubsub topic.
+Existing subscriptions will not be closed, but they will stop receiving messages.
+
+**Parameters**
+
+1. `int nodeId`: The node identifier obtained from a successful execution of [`waku_new`](#extern-char-waku_newchar-configjson).
+2. `char* pusubTopic`: Pubsub topic to unsubscribe from.
+  If `NULL`, unsubscribes from the default pubsub topic.
+
+**Returns**
+
+A [`JsonResponse`](#jsonresponse).
+If the execution is successful, the `result` field is set to `null`.
+
+For example:
+
+```json
+{
+   "result": null
+}
+```
+
+## Utils
+
+### `extern char* waku_encode_payload(char* data, char* keyType, char* key, char* signingKey, int version)`
```

Comment:
e.g. `extern char* waku_private_publish(int nodeId, char* messageJson, char* pubsubTopic, int timeoutMs, char* keyType, char* key, char* signingKey)`
where `messageJson` would now be an API construct that does not include `version` in the message.
Granted there may be cleaner ways, but it depends on whether we want API users to have to construct their own encrypted messages (and trust that they use the correct `version`) or to have the API take care of that for them behind the scenes.

Reply:
Interesting point. I was going to write that I opted for a similar API to create Waku Messages in js-waku but actually, to get Waku Messages the API decrypts for you.

I like the idea of simplifying the API. I would need to play with js-waku to have a better feel and opinion on the matter.

How do you intend to evolve the JSON RPC API for version 2?

One thought is the API is currently named `private`. I think it may be easier to call it `version_1` so that we don't have to re-use/break compability for `version_2`.

Or maybe one more abstraction layer would be ideal:

- `waku_relay_publish(message)`: version 0
- `waku_relay_publish_enc_asymmetric(message, publicKey, optionalSigningKey)`: version 1 asymmetric
- `waku_relay_publish_enc_symmetric(message, secretKey, optionalSigningKey)`: version 1 symmetric

You'd want to expose one API that takes an `encryption type` in argument if you intend to programmatically switch between both encryption types. I don't think we would expect an API consumer to do that.

Reply:
@richard-ramos WDYT?

Reply:
I ended up implementing the following functions:
```c
extern char* waku_lightpush_publish(char* messageJSON, char* topic, char* peerID, int ms);
extern char* waku_lightpush_publish_enc_asymmetric(char* messageJSON, char* topic, char* peerID, char* publicKey,     char* optionalSigningKey, int ms);
extern char* waku_lightpush_publish_enc_symmetric(char* messageJSON, char* topic, char* peerID, char* symmetricKey    , char* optionalSigningKey, int ms);

extern char* waku_relay_publish(char* messageJSON, char* topic, int ms);
extern char* waku_relay_publish_enc_asymmetric(char* messageJSON, char* topic, char* publicKey, char* optionalSign    ingKey, int ms);
extern char* waku_relay_publish_enc_symmetric(char* messageJSON, char* topic, char* symmetricKey, char* optionalSi    gningKey, int ms);

extern char* waku_decode_symmetric(char* messageJSON, char* symmetricKey);
extern char* waku_decode_asymmetric(char* messageJSON, char* privateKey);
```

In the case of lightpush, notice there's a peerID parameter. This is optional: if `null` is specified, go-waku auto select a peer. Otherwise, it will use the specified peerID for publishing the message.

Reply:
@richard-ramos can you please confirm that 
- `extern char* waku_encode_payload(char* data, char* keyType, char* key, char* signingKey, int version)`
- `extern char* waku_decode_payload(char* payload, char* keyType, char* key, int version)`
Can be removed then?

Reply:
We have currently defined a `JsonMessage` as follows:

```ts
{
    payload: string;
    contentTopic: string;
    version: number;
    timestamp: number;
}
```

However, we now have different API for version 0 and version 1.

I guess that 
- `waku_relay_publish`
- `waku_relay_publish_enc_asymmetric`
- `waku_relay_publish_enc_symmetric`

Do not use the `version` field of the `JsonMessage`, do they?

However, I guess that a `JsonMessage` received from a subscription do contains the `version` field and if the field is `1` then `waku_decode_*` should be used, right?

Reply:
> @richard-ramos can you please confirm that
> 
> * `extern char* waku_encode_payload(char* data, char* keyType, char* key, char* signingKey, int version)`
> * `extern char* waku_decode_payload(char* payload, char* keyType, char* key, int version)`
>   Can be removed then?

Yes, these can be removed. The list of functions i currently support are these:

- `waku_relay_publish`

for encrypting messages:
- `waku_relay_publish_enc_asymmetric`
- `waku_relay_publish_enc_symmetric`

for decrypting messages:
- `waku_dec_symmetric(messageJSON *C.char, symmetricKey *C.char) *C.char`
- `waku_dec_asymmetric(messageJSON *C.char, privateKey *C.char) *C.char`

Reply:
> We have currently defined a `JsonMessage` as follows:
> 
> ```ts
> {
>     payload: string;
>     contentTopic: string;
>     version: number;
>     timestamp: number;
> }
> ```
> 
> However, we now have different API for version 0 and version 1.
> 
> I guess that
> 
> * `waku_relay_publish`
> * `waku_relay_publish_enc_asymmetric`
> * `waku_relay_publish_enc_symmetric`
> 
> Do not use the `version` field of the `JsonMessage`, do they?
> 
> However, I guess that a `JsonMessage` received from a subscription do contains the `version` field and if the field is `1` then `waku_decode_*` should be used, right?

This is correct, for sending a message, the `version` attribute in a`messageJSON` is ignored and overwritten:
- `waku_relay_publish`   will have version 0
- `waku_relay_publish_enc_asymmetric` will have version 1
- `waku_relay_publish_enc_symmetric` will have version 1

The same logic applies to `waku_lightpush_*` functions.
For decrypting, the functions that require a key to decrypt the message expect that it contains version `1`.